src/inc/engine/render_component.h

```diff
@@ -35,7 +36,8 @@ class RenderComponent
 
   virtual void calculate_destination() = 0;
 
-  template<typename Impl>
+  template<typename Impl, typename = typename std::enable_if_t<
```

Comment:
Using enable if here to prevent compilation for `as_implementation` if `Impl` is not a derived class of `RenderComponent`. Still not happy with the `dynamic_cast`, but at least it's a bit guarded now.